ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/GenesisDensityGovernor.hs

```diff
@@ -0,0 +1,198 @@
+          -- "prompt reprocessing" in ChainSel).
+
+          theirTips <-
+            flip Map.traverseWithKey competingFrags $ \peer _ ->
+              cschTheirTip (handles Map.! peer)
+
+{-
+
+            |--------|
+
+    frag1: A - B - C - D - ...            <- exact
+             \
+    frag2:     E           claimed tip: E <- exact
+
+-}
+
+          let densityBounds = Map.fromList $ do
+                (peer, (frag, theirTip)) <- Map.toList $
+                    Map.intersectionWith (,) competingFrags theirTips
+                let lowerBound = fromIntegral $ AF.length frag
+                    startSlot = succWithOrigin $ AF.headSlot frag
+                    claimedTip = succWithOrigin $ getTipSlotNo theirTip
+                    hasBlockAfterGenesisWindow =
+                        succWithOrigin (AF.headSlot (candidateSuffixes Map.! peer))
+                      > endOfGenesisWindow
+                    endSlot
+                      | hasBlockAfterGenesisWindow = startSlot
+                      | otherwise = endOfGenesisWindow `min` claimedTip
+                    upperBound =
+                        lowerBound
+                      + unSlotNo (intervalLength startSlot endSlot)
+                pure (peer, (lowerBound, upperBound))
+
+              lowerBounds = Map.map fst densityBounds
+              upperBounds = Map.map snd densityBounds
+
+              killPeers
+                | Just maxLowerBound <- maximumMay $ Map.elems lowerBounds
+                , maxLowerBound > k
+                , let losingPeers =
+                        -- TODO: can probably already kill when <=
+                        -- maxLowerBound, might need to include case of a single
+                        -- peer with exactly known density.
+                        Map.keys $ Map.filter (< maxLowerBound) upperBounds
+                = for_ losingPeers $ \peer -> cschKill (handles Map.! peer)
+                | otherwise = pure ()
+
+          pure (killPeers, newCandidateTips)
+
+        killPeers
+
+        go newCandidateTips
+
+    SecurityParam k = configSecurityParam cfg
+
+    -- Length of the interval [a,b[ (lower bound inclusive, upper bound
+    -- exclusive).
+    intervalLength a b
+      | a <= b    = b - a
+      | otherwise = 0
+
+{-
+
+more TODO:
+
+ - we don't yet check that the header fragments contain no blocks from the
+   future (will likely be fixed by efforts not directly related to Genesis)
+
+ - idea by Torsten: seems like it is enough to run this logic whenever we add a
+   block to the ChainDB (instead of having yet another background thread)
+
+-}
+
+-- | Strip the common prefix of all fragments in a map.
+--
+-- PRECONDITION: all fragments have the same anchor
+--
+-- TODO: move elsewhere
+stripCommonPrefix ::
+     forall a blk.
+     (Ord a, HasHeader blk)
+  => Map a (AnchoredFragment blk)
+  -> (AnchoredFragment blk, Map a (AnchoredFragment blk))
+stripCommonPrefix frags = case traverse unconsLast frags of
```

Comment:
> Would it be simpler and sufficient to do it in two high-level steps: one fold to intersect them all, and then a map to split each against the intersection?

That sounds nicer than a direct recursive approach; I went with that (despite requiring two "obviously" unreachable `error` calls). I also realised that it can be generalised for arbitrary `f` with `Functor` and `Foldable`, not just `Map a`; I went with that as it stresses that we don't use anything special for `Map`s here; only downside is that we implicitly use the lazy implementations for `Map` this way :sob:.

> Also, when the set is empty, I suspect you should return the current immutable tip (which is in-scope at the call site), not necessarily `AnchorGenesis`.

Yeah, one further approach would be to use some non-empty type to avoid having to make decisions for the empty case here.